src/core/lib/promise/try_concurrently.h

```diff
@@ -0,0 +1,326 @@
+      : done_bits_(0),
+        pre_main_(std::move(other.pre_main_)),
+        post_main_(std::move(other.post_main_)) {
+    GPR_DEBUG_ASSERT(other.done_bits_ == 0);
+    other.done_bits_ = HelperBits();
+    Construct(&main_, std::move(other.main_));
+  }
+  TryConcurrently& operator=(TryConcurrently&& other) noexcept {
+    GPR_DEBUG_ASSERT(other.done_bits_ == 0);
+    done_bits_ = 0;
+    other.done_bits_ = HelperBits();
+    pre_main_ = std::move(other.pre_main_);
+    post_main_ = std::move(other.post_main_);
+    Construct(&main_, std::move(other.main_));
+    return *this;
+  }
+
+  ~TryConcurrently() {
+    if (done_bits_ & 1) {
+      Destruct(&result_);
+    } else {
+      Destruct(&main_);
+    }
+    pre_main_.template Destroy<1>(done_bits_);
+    post_main_.template Destroy<1 + PreMain::Size()>(done_bits_);
+  }
+
+  using Result =
+      typename PollTraits<decltype(std::declval<PromiseLike<Main>>()())>::Type;
+
+  Poll<Result> operator()() {
+    auto r = pre_main_.template Run<Result, 1>(done_bits_);
+    if (auto* status = absl::get_if<Result>(&r)) {
+      GPR_DEBUG_ASSERT(!IsStatusOk(*status));
+      return std::move(*status);
+    }
+    if ((done_bits_ & 1) == 0) {
+      auto p = main_();
+      if (auto* status = absl::get_if<kPollReadyIdx>(&p)) {
+        done_bits_ |= 1;
+        Destruct(&main_);
+        Construct(&result_, std::move(*status));
+      }
+    }
+    r = post_main_.template Run<Result, 1 + PreMain::Size()>(done_bits_);
+    if (auto* status = absl::get_if<Result>(&r)) {
+      GPR_DEBUG_ASSERT(!IsStatusOk(*status));
+      return std::move(*status);
+    }
+    if ((done_bits_ & NecessaryBits()) == NecessaryBits()) {
+      return std::move(result_);
+    }
+    return Pending{};
+  }
+
+  template <typename P>
+  auto NecessaryPush(P p);
+  template <typename P>
+  auto NecessaryPull(P p);
+  template <typename P>
+  auto Push(P p);
+  template <typename P>
+  auto Pull(P p);
+
+ private:
+  constexpr uint8_t NecessaryBits() {
+    return 1 | (PreMain::NecessaryBits() << 1) |
+           (PostMain::NecessaryBits() << (1 + PreMain::Size()));
+  }
+  constexpr uint8_t AllBits() {
+    return (1 << (1 + PreMain::Size() + PostMain::Size())) - 1;
+  }
+  constexpr uint8_t HelperBits() { return AllBits() ^ 1; }
+  uint8_t done_bits_;
+  PreMain pre_main_;
+  union {
+    PromiseLike<Main> main_;
+    Result result_;
+  };
+  PostMain post_main_;
+};
+
+template <typename Main, typename PreMain, typename PostMain>
+auto MakeTryConcurrently(Main&& main, PreMain&& pre_main,
+                         PostMain&& post_main) {
+  return TryConcurrently<Main, PreMain, PostMain>(
+      std::forward<Main>(main), std::forward<PreMain>(pre_main),
+      std::forward<PostMain>(post_main));
+}
+
+template <typename Main, typename PreMain, typename PostMain>
+template <typename P>
+auto TryConcurrently<Main, PreMain, PostMain>::NecessaryPush(P p) {
+  GPR_DEBUG_ASSERT(done_bits_ == 0);
+  done_bits_ = HelperBits();
```

Comment:
Does this simply set done_bits_ to 1 ?

Reply:
no... hopefully the new comments help: it's all bits except 1 (which corresponds to the main promise which we do not force complete here)

Reply:
Thanks the comments made it clearer